cli/lsp/logging.rs

```diff
@@ -1,13 +1,65 @@
 // Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
 
+use deno_core::parking_lot::Mutex;
+use deno_runtime::tokio_util::create_basic_runtime;
+use std::fs;
+use std::io::prelude::*;
+use std::path::PathBuf;
 use std::sync::atomic::AtomicBool;
 use std::sync::atomic::AtomicUsize;
 use std::sync::atomic::Ordering;
+use std::thread;
 
 static LSP_DEBUG_FLAG: AtomicBool = AtomicBool::new(false);
 static LSP_LOG_LEVEL: AtomicUsize = AtomicUsize::new(log::Level::Info as usize);
 static LSP_WARN_LEVEL: AtomicUsize =
   AtomicUsize::new(log::Level::Warn as usize);
+static LOG_FILE: Mutex<Option<LogFile>> = Mutex::new(None);
+
+pub struct LogFile {
+  path: PathBuf,
+  buffer: String,
+}
+
+impl LogFile {
+  pub fn log(&mut self, s: &str) {
+    self.buffer.push_str(s);
+    self.buffer.push('\n');
+  }
+
+  fn commit(&mut self) {
+    if !self.buffer.is_empty() {
+      if let Ok(file) = fs::OpenOptions::new().append(true).open(&self.path) {
+        if write!(&file, "{}", &self.buffer).is_ok() {
+          self.buffer.clear();
+        }
+      }
+    }
+  }
+}
+
+pub fn init_log_file(path: PathBuf) {
+  fs::write(&path, "").ok();
+  *LOG_FILE.lock() = Some(LogFile {
+    path,
+    buffer: String::with_capacity(1024),
+  });
+  thread::spawn(|| {
+    let runtime = create_basic_runtime();
+    runtime.block_on(async {
+      loop {
+        thread::sleep(std::time::Duration::from_secs(1));
+        if let Some(log_file) = &mut *LOG_FILE.lock() {
+          log_file.commit();
+        }
+      }
+    });
+  });
+}
+
+pub fn log_file() -> &'static Mutex<Option<LogFile>> {
```

Comment:
Let's rewrite this somehow to not publicly expose a mutex across modules (even though it's only used in the code below).